src/script-events.c

```diff
@@ -0,0 +1,343 @@
+
+    SET_OFFSET(CRC_DENTRY_D_PARENT);
+    u32 parent = *(u32 *)offset; // offset of d_parent
+
+    u32 *_to_skip = (u32 *)bpf_map_lookup_elem(&read_path_skip, &to_skip);
+    if (_to_skip)
+    {
+        to_skip = *_to_skip;
+    }
+
+    if (to_skip != 0)
+    {
+        SKIP_PATH_N(150);
+    }
+
+Send:
+    SEND_PATH_N(10);
+
+    // update to_skip, reuse skipped as index by resetting it
+    skipped = 0;
+    bpf_map_update_elem(&read_path_skip, &skipped, &to_skip, BPF_ANY);
+    // tail call back in
+    bpf_tail_call(ctx, &tail_call_table, 0);
+
+Skip:
+    skipped = 0;
+    bpf_map_delete_elem(&read_path_skip, &skipped);
+    ev->id = id;
+    return ev;
+}
+
+SEC("kretprobe/ret_script_load")
+int kretprobe__ret_script_load(struct pt_regs *ctx)
+{
+    char br = 0;
+    u32 id = 0;
+    u32 count = 0;
+    void **bprmp = NULL;
+    void *str = NULL;
+    u64 loaded = CRC_LOADED;
+    unsigned char *bprm = NULL;
+
+    // Make sure the function succeeded
+    int ret = PT_REGS_RC(ctx);
+    if (ret < 0)
+    {
+        return 0;
+    }
+
+    // Just to be safe 0 out the structs
+    telemetry_event_t ev;
+    __builtin_memset(&ev, 0, sizeof(ev));
+
+    // Initialize some of the telemetry event
+    id = bpf_get_prandom_u32();
+    ev.id = id;
+    ev.done = 0;
+    ev.telemetry_type = TE_SCRIPT;
+    ev.u.script_info.mono_ns = bpf_ktime_get_ns();
+
+    // Get Process data and set pid and comm string
+    ev.u.script_info.process.pid = (u32)bpf_get_current_pid_tgid();
+    bpf_get_current_comm(ev.u.script_info.process.comm, sizeof(ev.u.script_info.process.comm));
+
+    // Verify that the offsets are loaded
+    loaded = (u64)bpf_map_lookup_elem(&offsets, &loaded);
+    if (!loaded)
+    {
+        return 0;
+    }
+
+    // Get the stored pointer to the linux_binprm structure and clear it out of the map
+    bprmp = bpf_map_lookup_elem(&load_script_map, &ev.u.script_info.process.pid);
+    if (NULL == bprmp)
+    {
+        return 0;
+    }
+    bpf_map_delete_elem(&load_script_map, &ev.u.script_info.process.pid);
+
+    // Make sure it isn't NULL
+    bprm = (unsigned char *)*bprmp;
+    if (NULL == bprm)
+    {
+        return 0;
+    }
+
+    // Create the initial event in the cache
+    bpf_printk("Creating the initial event\n");
+    push_telemetry_event(ctx, &ev);
+
+    // Get filename
+    ret = read_value(bprm, CRC_BPRM_FILENAME, &str, sizeof(str));
+    if (ret < 0)
+    {
+        return 0;
+    }
+
+    // Read the filename
+    count = bpf_probe_read_str(&ev.u.v.value, sizeof(ev.u.v.value), str);
+    str += sizeof(ev.u.v.value) - 1;
+    bpf_printk("Count: %d\n", count);
+    br = ev.u.v.value[0];
+
+    // Output the path
+    ev.id = id;
+    ev.done = 0;
+    ev.telemetry_type = TE_CHAR_STR;
+    push_telemetry_event(ctx, &ev);
+    if (count >= sizeof(ev.u.v.value))
+    {
+        bpf_printk("count == to value size. Reading more...\n");
+        READ_CHAR_STR_N(10)
```

Comment:
Is this trying to get the whole path in the first 10 iterations and, if it can't get it, we go into the tail-call loop in `enter_script_load`?

Reply:
Kind of yea. I upped it to 29 in the latest version (4096/144) or (PATH_MAX / VALUE_SIZE). There is a filename filed in the linux_binprm structure. This contains whatever the user used to run the script. If this is an absolute path then all we need to do is read this field and we don't need to call enter_script_load since we don't need the cwd.